nvflare/app_common/ccwf/client_controller_executor.py

```diff
@@ -0,0 +1,382 @@
+    ):
+        """
+        ClientControllerExecutor for running controllers on client-side using WFCommClient.
+
+        Args:
+            controller_id_list: List of controller ids, used in order.
+            task_name_prefix: prefix of task names. All CCWF task names are prefixed with this.
+            persistor_id: ID of the persistor component
+            final_result_ack_timeout: timeout for sending final result to participating clients
+            max_task_timeout: Maximum task timeout for Controllers using WFCommClient when `task.timeout` is set to 0. Defaults to 3600.
+        """
+        check_number_range("final_result_ack_timeout", final_result_ack_timeout, min_value=1.0)
+
+        Executor.__init__(self)
+        self.controller_id_list = controller_id_list
+        self.task_name_prefix = task_name_prefix
+        self.persistor_id = persistor_id
+        self.final_result_ack_timeout = final_result_ack_timeout
+        self.max_task_timeout = max_task_timeout
+
+        self.start_task_name = make_task_name(task_name_prefix, Constant.BASENAME_START)
+        self.configure_task_name = make_task_name(task_name_prefix, Constant.BASENAME_CONFIG)
+        self.report_final_result_task_name = make_task_name(task_name_prefix, Constant.BASENAME_REPORT_FINAL_RESULT)
+
+        self.persistor = None
+
+        self.current_status = StatusReport()
+        self.last_status_report_time = time.time()  # time of last status report to server
+        self.config = None
+        self.workflow_id = None
+        self.finalize_lock = threading.Lock()
+
+        self.asked_to_stop = False
+        self.status_lock = threading.Lock()
+        self.engine = None
+        self.me = None
+        self.is_starting_client = False
+        self.workflow_done = False
+        self.fatal_system_error = False
+
+    def get_config_prop(self, name: str, default=None):
+        """
+        Get a specified config property.
+        Args:
+            name: name of the property
+            default: default value to return if the property is not defined.
+        Returns:
+        """
+        if not self.config:
+            return default
+        return self.config.get(name, default)
+
+    def start_run(self, fl_ctx: FLContext):
+        self.engine = fl_ctx.get_engine()
+        if not self.engine:
+            self.system_panic("no engine", fl_ctx)
+            return
+
+        runner = fl_ctx.get_prop(FLContextKey.RUNNER)
+        if not runner:
+            self.system_panic("no client runner", fl_ctx)
+            return
+
+        self.me = fl_ctx.get_identity_name()
+
+        self.persistor = self.engine.get_component(self.persistor_id)
+        if not isinstance(self.persistor, LearnablePersistor):
+            self.log_warning(
+                fl_ctx, f"Persistor {self.persistor_id} must be a Persistor instance, but got {type(self.persistor)}"
+            )
+            self.persistor = None
```

Comment:
Does this persistor have to be a `LearnablePersistor`?
If this is what it is as before then its fine for this PR, we can think in the future how to improve this class.

Reply:
This is how we checked persistors previously, will leave in this PR but agree that this can be improve in the future